### PR TITLE
fix: missing common maven transitive dependencies

### LINKED ISSUE #44: Missing full path for all transitive dependencies
### What happened?

When 2 dependencies have the same transitive dependency only one of the paths is fully represented in the SBOM.

e.g.
```
a -> b
b -> c
c -> d
a -> c
c -> d //missing
```

### Please provide runtime information.

Any environment

### Relevant log output

_No response_

### How can this issue be reproduced?

Using this example pom:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>
  <groupId>pom-with-deps-no-ignore</groupId>
  <artifactId>pom-with-dependency-not-ignored-for-tests</artifactId>
  <version>0.0.1</version>
  <dependencies>
    <dependency>
      <groupId>com.fasterxml.jackson.core</groupId>
      <artifactId>jackson-databind</artifactId>
      <version>2.14.1</version>
    </dependency>
    <dependency>
      <groupId>org.infinispan.protostream</groupId>
      <artifactId>protostream</artifactId>
      <version>4.6.4.Final</version>
    </dependency>
  </dependencies>
</project>
```

The resulting dot file after running the dependency:tree command `mvn -q dependency:tree -DoutputType=dot -DoutputFile=dependency.txt -f pom.xml`

```
digraph "pom-with-deps-no-ignore:pom-with-dependency-not-ignored-for-tests:jar:0.0.1" { 
	"pom-with-deps-no-ignore:pom-with-dependency-not-ignored-for-tests:jar:0.0.1" -> "com.fasterxml.jackson.core:jackson-databind:jar:2.14.1:compile" ; 
	"pom-with-deps-no-ignore:pom-with-dependency-not-ignored-for-tests:jar:0.0.1" -> "org.infinispan.protostream:protostream:jar:4.6.4.Final:compile" ; 
	"com.fasterxml.jackson.core:jackson-databind:jar:2.14.1:compile" -> "com.fasterxml.jackson.core:jackson-annotations:jar:2.14.1:compile" ; 
	"com.fasterxml.jackson.core:jackson-databind:jar:2.14.1:compile" -> "com.fasterxml.jackson.core:jackson-core:jar:2.14.1:compile" ; 
	"org.infinispan.protostream:protostream:jar:4.6.4.Final:compile" -> "org.jboss.logging:jboss-logging:jar:3.5.0.Final:compile" ; 
	"org.infinispan.protostream:protostream:jar:4.6.4.Final:compile" -> "com.squareup:protoparser:jar:4.0.3:compile" ; 
	"org.infinispan.protostream:protostream:jar:4.6.4.Final:compile" -> "org.javassist:javassist:jar:3.29.1-GA:compile" ; 
 } 
```

However when using the `verbose` flag (only available in `text` outputType):

```
pom-with-deps-no-ignore:pom-with-dependency-not-ignored-for-tests:jar:0.0.1
+- com.fasterxml.jackson.core:jackson-databind:jar:2.14.1:compile
|  +- com.fasterxml.jackson.core:jackson-annotations:jar:2.14.1:compile
|  \- com.fasterxml.jackson.core:jackson-core:jar:2.14.1:compile
\- org.infinispan.protostream:protostream:jar:4.6.4.Final:compile
   +- org.jboss.logging:jboss-logging:jar:3.5.0.Final:compile
   +- com.squareup:protoparser:jar:4.0.3:compile
   +- (com.fasterxml.jackson.core:jackson-core:jar:2.14.1:compile - omitted for duplicate)
   \- org.javassist:javassist:jar:3.29.1-GA:compile
```

### PR DESCRIPTION
**Pull request by ruromero**
## Description

The `dependency:tree` plugin omits duplicates but when using the `verbose` flag the omitted are shown. However the plugin only yields the `text` type in that case.

I have added the `verbose` flag to the `mvn dependency:tree` command and changed how the text file is processed with the `text` format.

**Related issue (if any):** 
fixes #44 

## Checklist

- [x] I have followed this repository's contributing guidelines.
- [x] I will adhere to the project's code of conduct.

## Additional information

Related to https://github.com/RHEcosystemAppEng/exhort/pull/135

### CODE REVIEW

**Comment by ruromero** [comment]
Do not merge, it seems the current solution does not load the full sub-tree, only the omitted dependency.

File: test/providers/java_maven.test.js
Comment (zvigrinberg) [inline test/providers/java_maven.test.js:29]: @ruromero Need to rename it to "pom_deps_with_no_ignore_common_paths" ( actual directory name) , because test is failing as a result of this typo

**Review by zvigrinberg — COMMENTED** [review]
@ruromero I Saw that there was a failure on test - RangeError: Maximum call stack size exceeded - couple of times actually -  please check it out at : https://github.com/RHEcosystemAppEng/exhort-javascript-api/actions/runs/6051817204/job/16424074479?pr=45#step:7:107.

It seems that it's either the halting condition of the recursion is not met on such case, or it's using too much space complexity in this recursion on big trees 

I'll try to figure out why it's happening for part of the tests ( while working on other tests).

Can you please take a look?

Thanks.

File: test/providers/java_maven.test.js
Comment (zvigrinberg) [reply test/providers/java_maven.test.js]: I added this recursion stop/halting condition to the code, because there are a lot of cases that there is a dependency with no continuing dependencies(even no direct ones) and it continues with no lines at all , or one empty lines ( the added halting condition)

```javascript
// exists condition
	if(lines.length === 0) {
		return
	}

// new condition in addition.
	if((lines.length === 1 && lines[0].trim() === ""))
	{
		return
	}
```

This one seems to solve majority of the failures, but we still have 2, one with mismatch of the content of sbom ( expected and actual) , and the other one that it inserted into infinity loop

Comment (zvigrinberg) [reply test/providers/java_maven.test.js]: @ruromero So after invetigation that i've made, it turns out that setting on the verbose flag on big pom.xml files (with or without parents) causing the mvn:dependency plugin + goal to enter infinite loop, that's why the test already run for 30 minutes without progressing..

I Tested it myself , and it's indeed stuck on it, so eventually i gave up after couple of minutes
```shell
[zgrinber@zgrinber poms_deps_with_no_ignore_long]$ date
**Sun Sep  3 14:43:40 IDT 2023**
[zgrinber@zgrinber poms_deps_with_no_ignore_long]$ mvn  dependency:tree -DoutputType=text  -DoutputFile=file.output -Dverbose
[INFO] Scanning for projects...
[INFO] 
[INFO] --------------< com.redhat.examples:custom-metrics-demo >---------------
[INFO] Building custom-metrics-demo 0.0.1-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-dependency-plugin:3.2.0:tree (default-cli) @ custom-metrics-demo ---
[WARNING] The artifact xml-apis:xml-apis:jar:2.0.2 has been relocated to xml-apis:xml-apis:jar:1.0.b2
[WARNING] The artifact ehcache:ehcache:jar:1.2 has been relocated to net.sf.ehcache:ehcache:jar:1.2
[WARNING] The artifact xerces:xerces:jar:2.4.0 has been relocated to xerces:xercesImpl:jar:2.4.0
[WARNING] The artifact bouncycastle:bctsp-jdk14:jar:138 has been relocated to org.bouncycastle:bctsp-jdk14:jar:1.38
[WARNING] The artifact com.lowagie:itext:jar:4.2.0 has been relocated to com.itextpdf:itextpdf:jar:5.5.6: After release 2.1.7, iText moved from the MPLicense to the AGPLicense.
			The groupId changed from com.lowagie to com.itextpdf and the artifactId from itext to itextpdf.
			See http://itextpdf.com/functionalitycomparison for more information.
[WARNING] The artifact activation:activation:jar:1.0.2 has been relocated to javax.activation:activation:jar:1.0.2
[WARNING] The artifact jboss:javassist:jar:3.3.ga has been relocated to javassist:javassist:jar:3.3.GA
[WARNING] The artifact poi:poi:jar:2.5.1 has been relocated to poi:poi:jar:2.5.1-final-20040804
[WARNING] The artifact servletapi:servletapi:jar:2.3 has been relocated to javax.servlet:servlet-api:jar:2.3
[WARNING] The artifact javacc:javacc:jar:3.2 has been relocated to net.java.dev.javacc:javacc:jar:3.2
[WARNING] The artifact xstream:xstream:jar:1.2 has been relocated to com.thoughtworks.xstream:xstream:jar:1.2
[WARNING] The artifact xerces:xerces:jar:2.0.2 has been relocated to xerces:xercesImpl:jar:2.0.2
[WARNING] The artifact itext:itext:jar:1.3.1 has been relocated to com.lowagie:itext:jar:1.3.1
[WARNING] The artifact itext:itext:jar:1.3 has been relocated to com.lowagie:itext:jar:1.3
[WARNING] The artifact jstl:jstl:jar:1.0.6 has been relocated to javax.servlet:jstl:jar:1.0.6
[WARNING] The artifact xerces:xerces:jar:2.2.1 has been relocated to xerces:xercesImpl:jar:2.2.1
[WARNING] The artifact sun-jaf:activation:jar:1.0.2 has been relocated to javax.activation:activation:jar:1.0.2: This file has been relocated based on maven recommendations: http://maven.apache.org/guides/mini/guide-coping-with-sun-jars.html
[WARNING] The artifact javax.measure:jsr-275:jar:1.0.0 has been relocated to javax.measure:jsr-275:jar:0.9.1
[WARNING] The artifact org.apache.commons:commons-io:jar:1.3.2 has been relocated to commons-io:commons-io:jar:1.3.2: https://issues.sonatype.org/browse/MVNCENTRAL-244
[WARNING] The artifact junit:junit-dep:jar:4.11-beta-1 has been relocated to junit:junit:jar:4.11-beta-1: The artifact junit:junit does not contain Hamcrest anymore but declares a dependency to Hamcrest. Thus, junit:junit-dep has become obsolete.
^C[zgrinber@zgrinber poms_deps_with_no_ignore_long]$ date
**Sun Sep  3 14:50:15 IDT 2023**
```

image

Comment (zvigrinberg) [reply test/providers/java_maven.test.js]: @ruromero it appears that the infinite loop is not happening anymore when running the latest version of maven dependency plugin:
```shell
mvn -q org.apache.maven.plugins:maven-dependency-plugin:3.6.0:tree -Dverbose -DoutputFile=output.file
```

File: src/providers/java_maven.js
Comment (zvigrinberg) [inline src/providers/java_maven.js:96]: I've Changed to this regex in order to parse dependencies without special characters in all cases.

File: src/providers/java_maven.js
Comment (zvigrinberg) [inline src/providers/java_maven.js:110]: Although it works ( because it's a primitive type) , i think we better change it to tripe =,  and we'll remove this linting warning as a bonus

**Review by zvigrinberg — APPROVED** [review]
@ruromero Good Job.
 As you could see, i pushed few fixes to your fork, so it will work correctly and all tests will succeeded.
I Approved it - please Merge whenever you are ready ( maybe you want to complete documentation or add something - linting warning for example - see comment below about that - anyway it's not critical at the moment) 
Thanks.

**Comment by ruromero** [comment]
Thanks @zvigrinberg let me validate the use case reported by QE before merging

**Review by zvigrinberg — APPROVED** [review]
lgtm

**Pull request merged**
